text/0162-refactoring-support.md

```diff
@@ -0,0 +1,586 @@
+there are no refactors to apply from this diff, we can go ahead and deploy the
+application. Again, the new state results from the merge of `H(E)` and
+`H(A)`:
+
+    H(E) = e1 ◄── e2 ◄── e3         
+                   ▲                  
+                   │                  
+                   └──── e4 ◄── e5*
+
+If there are refactors to be done, this is considered an error, because we can't
+guarantee that the refactor makes sense (let alone that this was the developer's
+intent).
+
+#### The future
+
+There is still some work to be done to prove this system works in practice. But
+assuming it does, we could use it to expand the scope to which the automatic
+refactoring applies. Consider the case in which you want to rename a certain
+resource and, at the same time, make some minor changes, such as adding or
+updating a couple of properties. This is another ambiguous case, because it's
+not clear what the intent is: update with rename, or replacement? But with the
+history system, we can detect such cases, interact with the developer, and store
+the decision in the history.
+
+Since this historical model contains all the information about the state of the
+stacks in an environment, it could also be used for other purposes. For example,
+development tools could use the history to provide a "time machine" feature,
+that allows developers to see the state of the infrastructure at any point in
+time. CloudFormation itself could build on that, and provide a way to roll back
+or forward to another state.
+
+Another problem that could be solved with the historical model is the infamous
+"deadly embrace", where a consumer stack depends on a producer stack via
+CloudFormation Exports, and you want to remove the use from the consumer. At the
+moment, customers have to use the `stack.exportValue(...)` method, and do two
+deployments. The history would give the CLI all the information it needs to do
+this without user intervention.
+
+Potentially, this could also help with drift resolution (or prevention), if
+CloudFormation itself starts using the history internally.
+
+### B. Equivalence between resources
+
+To detect which resources should be refactored, we need to indentify which
+resources have only changed their location, but have remained "the same", in
+some sense. This can be made precise by defining an [equivalence relation] on
+the set of resources.
+
+Before that, let's define a digest function, `d`:
+
+    d(resource) = hash(type + physicalId)                       , if physicalId is defined
+                = hash(type + properties + dependencies.map(d)) , otherwise
+
+where `hash` is a cryptographic hash function. In other words, if a resource has
+a physical ID, its type and physical ID uniquely identify that resource. So we
+compute the hash from these two fields. Otherwise, the hash is computed from its
+type, its own properties (that is, excluding properties that refer to other
```

Comment:
ah makes sense! so base on this, the digest of a lambda function with property log group referring to another resource will remain the same after refactoring, as the log group property is excluded in the calculation.

Reply:
The digest of the log group is included in the calculation. But since logical IDs are never part of the digest, we can freely change the logical ID of the log group without affecting the digests of the resources that depend on it.